lib/active_record/immutable.rb

```diff
@@ -0,0 +1,49 @@
+module ActiveRecord
+  module Immutable
+    extend ActiveSupport::Concern
+
+    FIELD_NAME_IDENTIFIER_IMMUTABLE = 'previous_version_id'.freeze
+
+    included do
+      def self.default_scope
+        where(%{
+          not exists (
+            select 1
+            from #{quoted_table_name} model
+            where model.#{FIELD_NAME_IDENTIFIER_IMMUTABLE} = #{quoted_table_name}.id
+          )
+        })
+      end
+
+      def _update_record(_attribute_names = self.attribute_names)
+        create_new_version
+      end
+
+      def create_new_version
+        clone = build_clone
+        clone.immutable_field_value = id
+        clone.save
+      end
+
+      def build_clone
+        dup
+      end
+
+      def next_version
+        self.class.unscoped.find_by(FIELD_NAME_IDENTIFIER_IMMUTABLE => id)
+      end
+
+      def previous_version
+        self.class.unscoped.find(immutable_field_value) if immutable_field_value.present?
+      end
+
+      def immutable_field_value
```

Comment:
What do you think about renaming this to `previous_version_id_value`?

Reply:
Actually, I'm thinking that we can just use the `previous_version_id` for now, what do you think?

Reply:
I think it's better `previous_version_id_value`, to avoid changes in the future.